ui/src/app/machines/components/ActionFormWrapper/CloneForm/CloneFormFields/SourceMachineSelect/SourceMachineSelect.tsx

```diff
@@ -0,0 +1,146 @@
+import { useEffect, useState } from "react";
+
+import { MainTable, SearchBox, Spinner } from "@canonical/react-components";
+import { highlightSubString } from "@canonical/react-components/dist/utils";
+import { useDispatch, useSelector } from "react-redux";
+
+import DoubleRow from "app/base/components/DoubleRow";
+import LabelledList from "app/base/components/LabelledList";
+import { actions as machineActions } from "app/store/machine";
+import machineSelectors from "app/store/machine/selectors";
+import type { Machine } from "app/store/machine/types";
+import type { RootState } from "app/store/root/types";
+
+type Props = {
+  source?: Machine["system_id"];
+  setSource: (id: Machine["system_id"]) => void;
+};
+
+const generateRows = (
+  machines: Machine[],
+  searchText: string,
+  onRowClick: (machine: Machine) => void
+) => {
+  const highlightedText = (text: string) => (
+    <span
+      dangerouslySetInnerHTML={{
+        __html: highlightSubString(text, searchText).text,
+      }}
+      data-test={`source-machine-${text}`}
+    />
+  );
+
+  return machines.map((machine) => ({
+    className: "source-machine-select__row",
+    columns: [
+      {
+        content: (
+          <DoubleRow
+            primary={highlightedText(machine.hostname)}
+            secondary={highlightedText(machine.system_id)}
+          />
+        ),
+      },
+      {
+        content: (
+          <DoubleRow
+            primary={machine.owner || "-"}
+            secondary={
+              machine.tags.length
+                ? highlightedText(machine.tags.join(", "))
+                : "-"
+            }
+          />
+        ),
+      },
+    ],
+    "data-test": "source-machine-row",
+    onClick: () => onRowClick(machine),
+  }));
+};
+
+export const SourceMachineSelect = ({
+  source,
+  setSource,
+}: Props): JSX.Element => {
+  const dispatch = useDispatch();
+  const [searchText, setSearchText] = useState("");
+  const unselectedMachines = useSelector(machineSelectors.unselected);
+  const selectedMachine = useSelector((state: RootState) =>
+    machineSelectors.getById(state, source)
+  );
+  const loaded = useSelector(machineSelectors.loaded);
+  // We filter by a subset of machine parameters rather than using the search
+  // selector, because the search selector will match parameters that aren't
+  // included in the clone source table.
+  const filteredMachines = unselectedMachines.filter(
```

Comment:
Does this need to also limit the number of machines?

Reply:
Oh yeah maybe. `MainTable` will automatically paginate at 50 but maybe we're not supposed to be paginating in this little sub-table thing. I'll clarify with Amy.